src/dmd/dstruct.d

```diff
@@ -606,39 +606,53 @@ extern (C++) class StructDeclaration : AggregateDeclaration
         return index < numArgTypes() ? (*argTypes.arguments)[index].type : null;
     }
 
-    final bool hasNonDisabledCtor()
-    {
-        static extern (C++) class HasNonDisabledCtorVisitor : Visitor
-        {
-            bool result;
-
-            this() {}
-
-            alias visit = Visitor.visit;
 
-            override void visit(CtorDeclaration cd)
-            {
-                if (!(cd.storage_class & STC.disable))
-                    result = true;
-            }
+    /***************************************
+     * Verifies whether the struct declaration has a
+     * constructor that is not a copy constructor.
+     * Optionally, it can check whether the struct
+     * declaration has a regular constructor, that
+     * is not disabled.
+     *
+     * Params:
+     *      checkDisabled = if the struct has a regular
+                            non-disabled constructor
+     * Returns:
+     *      true, if the struct has a regular (optionally,
+     *      not disabled) constructor, false otherwise.
+     */
+    final bool hasRegularCtor(bool checkDisabled = false)
+    {
+        if (!ctor)
+            return false;
 
-            override void visit(TemplateDeclaration td)
+        bool result;
+        overloadApply(ctor, (Dsymbol s)
+        {
+            if (auto td = s.isTemplateDeclaration())
             {
+                if (checkDisabled && td.onemember)
+                {
+                    if (auto ctorDecl = td.onemember.isCtorDeclaration())
```

Comment:
Is that check correct? Are we guaranteed that there will only be one member for ctor decls ?

Reply:
This is correct if your constructor is always going to be a templated function in the short hand form i.e. `this(..)(..)`. It will not work if you try:

```d
struct Bar
{
 
    template __ctor(A)
    {
        this(A a) {}   
    }
    
    uint a;
    string b;
}
Bar b = { a: 42, b: "hello" };
```
But I would argue that nobody should use that form. Using `__ctor` is risky and has problems.

Reply:
I was more thinking of something like this:
```D
struct Foo
{
    this(T)(T a) if (is(T == int))
    {

    }

    this(T)(T a) if (is(T == long))
    {

    }

    string data;
}
```

Also:
```D
struct Foo
{
    template Tmpl ()
    {
        this(T)(T a) if (is(T == int))
        {

        }

        this(T)(T a) if (is(T == long))
        {

        }
    }

    mixin Tmpl!();

    string data;
}
```

Although those might just produce an overloadset ?

Reply:
If we are going to get rid of bypassing disabled constructors in favor of default construction for struct initializers, then this discussion becomes moot.

Anyway, I just tested and the current implementation works fine for both of you examples. By works fine, I mean: if both overloads are disabled, struct initialization is still going to work by using the default constructor.